### ISSUE TITLE
V0.4.0 extension stops working

### ISSUE THREAD

**Issue by pmario** [pmario/file-backups#34]
Have fun!
mario
### PS: If you use it: [Support it!](https://www.paypal.me/PMarioJo)

**Comment by jeroppi** [comment]
Dear Mario

Thank you very much for creating File-Backups for Tiddlywiki.

I am having this weird issue: File-backups (0.4.0)  works perfectly for a while, but then all of a sudden it stops saving the backups.
If I am lucky, at some point in time it will work again, (but just for a short time: maybe for a day or for some hours) and after a while it breaks again and seems to stop working for good.

I cannot recognize any cause nor any patterns. The behavior seems erratic to me.

Windows 10 (update 2004), Firefox 77.0.1 (64-bit).

Firefox download options settings (please see screenshot):

I have created a folder (other than Windows' "Downloads") for Firefox downloads, because I want to have my wikis and backups saved on their own partition of my Notebook's SSD. Firefox seems to be OK with that: "File-Backups" works like a charm, saving the wikis, creating the backup folders and populating these with the backups. (Except when it stops working.)

I have Firefox set up to "always ask where to save files", but when "File-Backups" is working fine, the wikis are correctly saved and backed up without Firefox asking me anything.

At some point in time I then notice that "File-Backups" has stopped working because Firefox starts to ask me where to save the wiki. I then browse to the usual location, but Firefox says that the file already exists, and my only option is to overwrite it. And no backups are created in the backup folders.

Is this a known issue?
Do you know if I can I do something to stop this issue?

Yours truly
Jeronimo Miniño  


![Firefox_download_options](https://user-images.githubusercontent.com/45630863/86029722-7d534700-ba6e-11ea-9b29-726e91a4fd5c.JPG)

**Comment by pmario** [comment]
>I have Firefox set up to "always ask where to save files", but when "File-Backups" is working fine, the wikis are correctly saved and backed up without Firefox asking me anything.

That's a configuration I didn't test. .. So I'll have to have a closer look here.

**Comment by pmario** [comment]
Is your PC running 24/7 and FireFox open all the time?

**Comment by jeroppi** [comment]
No, my PC is not exactly running 24/7 with Firefox open all the time.
Every 2 or 3 days I switch off the PC for a  couple of hours. The rest of the time, I have FF open (editing TWs).

**Comment by jeroppi** [comment]
I don´t have the impression that this setting matters much. I have also
tried the setting where Firefox saves without asking me. What happens then
is:
a) when File-Backups works: Everything works fine
b) when File-Backups stops working: each time I save a wiki, Firefox saves
a new (numbered) copy of the wiki in the folder that I have designated for
saving purposes, and nothing is saved in the backup folders.

An hour ago all of a sudden File-Backups started working fine again. I have
no idea what triggered that. According to my experience, during the next
hours I expect it to stop working again.

Kind regards

**Comment by pmario** [comment]
How big is your wiki in MByte?

**Comment by pmario** [comment]
Do you edit several wikis at the same time? ... I'm asking, so I can "rebuild" a setting, that is similar to yours.

**Comment by pmario** [comment]
So you use the "Always ask, where to save" as an indicator, when fb fails. right?

**Comment by pmario** [comment]
@jerojerojeroppi ... I did rename the issue and tagged it as bug

**Comment by jeroppi** [comment]
Currently, my biggest wiki is 14 MB.

btw I expect these wikis to grow as I keep adding contents and
functionality for my distance courses.
Apropos "Functionality": pluginmäßig benutze ich z.Z. meist "Noteslam",
TiddlyTables und TiddlyMap.

**Comment by jeroppi** [comment]
No, I usually don't edit several wikis at the same time. I usually work
with only one wiki open at a time.

**Comment by jeroppi** [comment]
Haha! It now happens to come handy as an indicator, but that's not the reason why I set FF that way. I just like to be in control of my downloads.

BTW I  File-Backups is the only extension I have installed in FF. No plugins either. I very rarely open webpages in other FF tabs.
I use FF almost exclusively for editing wikis.

**Comment by pmario** [comment]
Hi, ... I did create an 18MByte TW and have it opened all the time. From time to time I'm switching to the tab and make some changes. I did create the _exact_ same directory structure you used in your screenshot. 

The only differences are, that FF has updated in the meantime to 78.0.1 and my Windows is: Pro 10 (1903)

I do use more FF plugins. ... 

I couldn't replicate the problem till now. ... So we will have to wait and see. 

In the meantime you'll have to have a close look at you "indicator" ... the browser download dialog. ... 

--------

What I'd like to know, if file-backup stopps working. ... Does it help if you do a "Tab reload" for the wiki with F5. ... 

BUT be sure you did save your changes!!!

**Comment by jeroppi** [comment]
Thank you very much for your efforts. Sorry for the delayed answer. Very
busy days.
I have been frequently uninstalling and reinstalling the extension when it
stops. After reinstalling, it works again for a while.
Tonight or tomorrow I will try your idea with tab reloading and I will let
you know if it helps.

**Comment by jeroppi** [comment]
Hi.
Tonight's feedback:
As my old backup folders were all messed up by my former various attempts
of coping with the backup failing,..
tonight I moved all that old mess to a quarantined folder and
I created a new wikis folder on Partition D, containing only the newest
version of each wiki.
Then I uninstalled Firefox completely.
Then I downloaded Firefox and reinstalled it, with no old settings
whatsoever.
I set Firefox updates to NOT run automatically and NOT make use of the
"Mozilla Maintenance Service" (Before I had Updates run automatically).
I set FF's downloads location to the Wikis Folder.
I set FF to ask every time before saving downloads (the "indicator")
I installed the File-Backups extension. (I checked to "allow extension to
run private... gobbldygock" (don't remember exactly)
I loaded a wiki and did a small edit.
I clicked the Tiddlywiki's icon to save.
FF asked me where to save it. I browsed to the newly created wikis folder.
FF told me the file existed there, and asked me to overwrite it. (So It
seems to me that the backup failed from the very beginning).
Then I tried F5 as you suggested.
Then I did some edit on the wiki.
Then I clicked on the Tiddlywiki's save icon.
This time it worked!. It saved silently!
I went to check the Wikis folder, and I found the following:
- All of mi wikis are there, as they should be.
- The wiki that I edited is correctly edited.
- The beacon html file is there.
- The twBackups folder has been created.
- Inside of twBackups folder, the subfolder for backups of that specific
wiki was properly created.
- Inside that wiki's backups folder there were two backup files:
- the one which ends in "...(2020-07-09T19-19-55-145Z).html"
- and one which ends in "...(B).html"

F5 seems to have done the trick!. I will keep on giving feedback.

**Comment by jeroppi** [comment]
Oh, you could not replicate the issue... Merkwürdig!

If you think it's a good idea, I can send you a copy of one of my wikis (after deleting sensible data). So we could see if the issue is particular to my wikis.

As soon as I can take some time, I will see if I can replicate the issue on my second PC. I will give feedback.

FWIW: 

The PC I am using now is a Huawei Matebook Pro model 2018 I bought new 3 months ago. It seems to run full ok, I have not had any issues whatsoever. i7-8550u, 16GB RAM, dedicated nvidia graphics.
Windows 10 Home, as of today: version 2004, OS Build 19041.329

Wired connections: 

- a second screen (HDMI) 
- high speed internet (ethernet)

Wireless connections:

- wifi (same network  as the wired one), 

Periferals:

- bluetooth: mouse, headset
- old smartphone used as a webcam, connected through wifi

I am also using "Mouse Without Borders" to share the mouse, the keyboard and the clipboard with a second (tablet-)PC through WiFi.

**Comment by pmario** [comment]
... OK .. I saw a weakness. ... If you install the AddOn AND the wiki is already open, the AddOn isn't active for already open tabs. .. So they have to be refreshed. ... I'm pretty sure, that's what you saw, with the new setup.

**Comment by pmario** [comment]
I did create a new issue #37. So it won't be lost.

**Comment by jeroppi** [comment]
Hi! 

Thank you for your kind message and for your continuous efforts in improving fb. Last time I wrote that at some point I "loaded" a wiki. Sorry: what I meant with "loaded" is that I "opened" the wiki in a tab. I thought I did it after installing the extension, but I'm not 100% sure. Chances are you're right. 

I had to spend a day working on something unrelated to the wikis, but in a couple of hours I will start editing wikis again. I will send you feedback.

**Comment by jeroppi** [comment]
Hi. 
Today (at last!) I have been able to identify one thing that consistently triggers the behavior of FilesBackup stopping to save a wiki correctly in my PC:

First thing, I must state that I never open a wiki from Firefox's menu "Open file". I ALWAYS open my wikis by going to the wikis' folder, right-clicking on the file's name and then clicking "open with" and "Firefox" in Windows Explorer's context menu.

Well, what I noticed today is that when Firefox is already open AND THEN I open a wiki, everything runs fine: The wiki opens in a Firefox tab, and everytime I edit a tiddler or I click on the save icon, the wiki seems to be saved/backed up properly. 

BUT if I open a wiki (from Windows Explorer's context menu, as I always do) when Firefox is NOT YET open, then Firefox opens with the wiki in a tab, but the first time I try to save the wiki, it will NOT be saved/backed-up properly. What happens is that my "indicator" kicks in (i.e. the browser download dialog opens).

I am curious if you can replicate this.

From now on I intend to make sure Firefox is open before I open any wiki, so I can test if that was the only cause of the problem.

**Comment by jeroppi** [comment]
By the way, I always have Firefox set to "restore previous session" at the browser's startup.

I tried some more things:

Experiment 1.
- uncheck the option "restore previous session" at browser's startup in settings.
- this didn't seem to have any influence on the observed behavior.

Experiment 2.
- Open Firefox
- Open a wiki from Windows Explorer context menu ("Open with" "Firefox")
- Close Firefox (without having closed the wiki's tab)
- Reopen Firefox
- Firefox restores the wiki's tab
- Try to save the wiki... it saved OK.

Experiment 3.
- Firefox is closed.
- I Open a wiki from Windows Explorer context menu ("Open with" "Firefox")
- Firefox opens with one tab. This tab contains the wiki.
- Try to save the wiki... It failed: I got the download window.
- I cancel the download window without having overwritten the wiki.
- (I can repeat the former two steps any number of times)
- Then I close Firefox (without having closed the wiki's tab)
- Next I open Firefox
- Firefox restores the wiki's tab
- I try to save the wiki... Now the wiki was saved/backed up properly (and it kept saving/backing-up properly during the whole browser session).

Experiment 4:
- First I do one of the two previous procedures, (1.) or (2.) (it will make no difference)
- Then I close the browser again (without having closed the wiki's tab)
- Then I go to the wikis' folder, right click on that same wiki and "Open with" "Firefox".
- This time Firefox opens with the same wiki in two tabs. (I suppose that the first tab is the one restored from the previous session, and that the second tab contains the instance of the wiki which I just opened by right-clicking.)
- As it opens, the browser shows the wiki in the second tab. There I try to save the wiki... and the saving fails: The browser download window opens instead.
- I cancel that browser download window without overwriting the wiki
- Then I go to the first window. There I get a big red warning about the wiki being open in another tab.
- Not caring about the big red warning, I proceed to save the Wiki... IT SAVES OK.
- I can jump back and forth, and get the same results every time: On the first tab (the one with the big red warning) the wiki saves OK everytime. But on the second tab the wiki cannot be saved/backed-up properly.

**Comment by pmario** [comment]
Which FF version do you use?

**Comment by pmario** [comment]
Forget the last post. I saw it in the OP

**Comment by pmario** [comment]
As soon as the "red banner" is shown, the AddOn is active. The check, if several tabs contain the same wiki is done by the AddOn.

**Comment by pmario** [comment]
Do you work with "FF Private Mode"?

**Comment by pmario** [comment]
Which exact Windows Version do you use: Win 10 Home / Pro / Enterprise ... ?

**Comment by pmario** [comment]
hmmm. ... I could replicate it 1 time, with a new VM after a restart. .... So there is an initialisation problem. ... I can't reproduce it reliably. ... But since you can, it has to be that way. 

I'll have to have a closer look, how I can improve the init behaviour. ... It'll need some time. ... 

I hope you have enough info, to avoid the problem ... But as said. I'll have a look, how I can improve it.

**Comment by jeroppi** [comment]
I am currently running Firefox 78.0.2 (64-bit). I have FF set to update
automatically.

**Comment by jeroppi** [comment]
No, I don't work with FF "private mode"

**Comment by jeroppi** [comment]
I use Windows 10 Home, Version 2004, OS build 19041.388

**Comment by jeroppi** [comment]
Thank you very much. Yes, I think now I can manage. 

When I get to make tests on other computers, I will tell you the results.

I think my "Experiment 4." is likely to give you some hints on  how the problem arises: 
I can now reliably create the situation where I have the same wiki open in two browser tabs simultaneously: In one tab the addon is working fine, while at the same time, on the other tab the addon is not working. And I can reliably predict whether the addow will work in a tab or not, depending on the procedure I used to open the wiki in that tab. I will try to make a video capture of this. That might be clearer than my convoluted explanations (and my English! LOL).

**Comment by pmario** [comment]
Your workflow is completely different to mine. In my setting I use FF as the default browser. So I only need a "double click" to open the wiki. ... So I never ever tested to open a wiki with "open with" from the explorer. 

So I'm pretty sure, it's an init-problem. ... Once I do have an "universal init" set up right, the problem should be gone .... hopefully. 

So it's not really necessary to create more test cases, that use up a lot of your time. .. I think I do have enough info.

**Comment by jeroppi** [comment]
OK, I see, great! Thanks a lot!
Btw, one reason why I love this addon is because I find its backup scheme outstanding. Other solutions just create backup folders that grow and grow with no limit.

**Comment by Skeeve** [comment]
Hi Mario.
Today I wanted to start a new TW for a colleague I am the vacation stand-in for.

The new-empty wiki didn't save and so I landed here.

First the environment:
MacBookPro 10.14.6
FireFox 78.0.2 (64-bit) (Default Browser)
File Backups 0.4.0
Mac uptime 7 days

When I noticed the new TW didn't save, I opened my old TW (via Bookmark) and it didn't save either.

I opened a TW via dragging the file to an empty tab - it didn't save

I read the bug info found here and closed FF.

I then opened by double-clicking my new empty TW. I could save. (Hooray) ;)

I then opened my old TW via Bookmark and could save that as well.

**Comment by pmario** [comment]
Superseded by #37, the narrowed-down init-on-already-open-tabs issue that was identified as the root cause in this thread (particularly @jerojerojeroppi's Experiment 3 and 4 reproduction). Consolidating further work there. Thanks to everyone who helped debug this here — the detailed step-by-step repros were invaluable.

**Comment by pmario** [comment]
Reopening per owner request.
